### ISSUE TITLE
Implementing common polynomials

### ISSUE THREAD

**Issue by mortenpi** [JuliaMath/Polynomials.jl#33]
It would nice if it there would be an easy way to construct some of the standard polynomials (Hermite, Jacobi etc).

There isn't actually that much code available at the moment. The currently existing packages that I was able find:
- https://github.com/pjabardo/Jacobi.jl -- seems maintained but only provides Jacobi polynomials
- https://github.com/daviddelaat/Orthopolys.jl -- provides a few more, but is broken and not in the package index

The question is whether it should be a separate package or would it make sense to have the code here? I don't think it warrants a separate package and having to hunt through different packages to find the necessary function is a headache. It would be nice if you could just do something like

``` julia
using Polynomials
p = hermite(3)
```

If seems reasonable to have them here then I would love to make some PRs at some point.

**Comment by jverzani** [comment]
I thought about putting this in, but does `FastGaussQuadrature` cover what you want? There you can get the nodes for several families of orthogonal polynomials and from there use `poly` to construct the polynomials.

**Comment by mortenpi** [comment]
The short answer is no really, since what I would like to have is a simple way of constructing these polynomials. So, irrespective of how it would be implemented, it would be nice to have a method like `foo(N, args) -> Poly(...)` for each of the common polynomials available for the user. Also, I think the `Polynomials` package is exactly the right place for such functions, so that the user does not have to hunt for a method, but finds it in the obvious place.

I took a quick look at the `FastGaussQuadrature` package and I see the following issues with recommending that:
* You would have to be familiar with Gaussian quadrature to use it. I am vaguely but, it is not entirely obvious to me how you should construct the polynomials from the quadrature numbers.
* You only get floating point coefficients that way and hence all kinds of rounding errors, whereas the true coefficients are integers.
* Not entirely sure about this, but I would guess that it is more efficient to calculate the coefficients using a recurrence relation, instead of going through the Gaussian quadrature calculation (even if `FastGaussQuadrature` is very fast).
 
Only after posting this issue I saw #18, which is kind of what I would like to see. However, I don't think that the API there is very good. I would say that `hermite(N)` is more clear than `Orthogonal.H(N)`, for example.

I would propose the following API:
* Since there are several ones with the same name (physicists'/probabilists' Hermite,  1st and 2nd Chebyshev, etc), then perhaps the best convention for naming the would to just copy Matlab/Mathematica, so that the function names would be
`hermiteH`, `chebyshevT`, `chebyshevU`, `gegenbauerC`, `jacobiP`, `laguerreL`, `legendreP`. However, if there is no ambiguity, then perhaps the last letter can be dropped, since it's ugly, e.g. to have just `jacobi`, `laguerre`, etc. And for the probabilists' Hermite I would suggest `hermiteHe`.
* Each function would have the following signatures:
  - `P(N::Integer) = Poly{Int}([...])`
  - `P(N::Integer, x::Symbol) -> Poly{Int}([...], x)`

If a polynomial has arguments (e.g. Jacobi), then the signature would be `P(N, arg1, ..., x)`.

I think I will start writing this up in code to make a PR, but any thoughts are welcome.

**Comment by jverzani** [comment]
This sounds a lot like the Jacobi package you link to. I guess the approach depends on what you need to do. For numerics or higher-orders, the FastGaussQuadrature is likely preferred, though not for symbolic computations. Just to illustrate:

```
## from Jacobi.jl
function legendre(n, sym::Symbol=:x)
    x = Poly([0//1,1//1],sym)
    p0 = Poly([1//1], sym)
    p1 = x
    if n==0
        return p0
    elseif n==1
        return p1
    end

    for i = 2:n
        p2 = ( (2i-1)*x*p1 - (i-1)*p0 ) / i
        p1, p0 = p2, p1
    end

    return p1
end


using FastGuassQuadrature
function flegendre(n, sym::Symbol=:x)
    ns, wts = gausslegendre(n)
    poly(ns, sym)
end


@time legendre(25) # 0.000765 seconds (957 allocations: 85.672 KB)
@time flegendre(25)#  0.000036 seconds (30 allocations: 3.953 KB)

@time flegendre(100) # 0.000044 seconds (20 allocations: 7.313 KB)
@time legendre(100)  # overflow
```

Personally, I would think a separate package would probably be best. (Partly due to a desire to keep this one relatively simple and contained.)

As for names, I think passing in different kinds through arguments is preferable to having separate names.

**Comment by jverzani** [comment]
There is WIP on this at https://github.com/jverzani/SpecialPolynomials.jl building on @mileslucas `ChebyshevT` example  in the rewrite of the backend  for  `Polynomials`  that was  just merged. If there is still interest, please make  comments on SpecialPolynomials
